File: algorithms/quick-sort.js
// https://en.wikipedia.org/wiki/Quicksort
export function quickSort(arr) {
  if (arr.length < 2) return arr;

  const pivotIndex = Math.floor(arr.length / 2);

  const lessThanPivot = [];
  const greaterThanPivot = [];

  for (let i = 0; i < arr.length; i++) {
    if (i === pivotIndex) continue;

    if (arr[i] > arr[pivotIndex]) {
      greaterThanPivot.push(arr[i]);
    } else {
      lessThanPivot.push(arr[i]);
    }
  }

  return quickSort(lessThanPivot).concat(arr[pivotIndex]).concat(quickSort(greaterThanPivot));
}

File: original-python-solutions/quick-sort.py
# O(n log n)
def quick_sort(arr):
  if len(arr) < 2:
    return arr

  pivot = arr[0]
  less_than_pivot = [i for i in arr[1:] if i <= pivot]
  greater_than_pivot = [i for i in arr[1:] if i > pivot]

  return quick_sort(less_than_pivot) + [pivot] + quick_sort(greater_than_pivot)

assert quick_sort([10, 5, 2, 3]) == [2, 3, 5, 10]
